### PR TITLE
feat: Added preSerialization hook and route option.

### LINKED ISSUE #1369: Add preSerialization hooks as a route-level hooks
## 🚀 Feature Proposal

Add preSerialization hooks as a route-level hook, called when reply.send() is called (just before response serialization is started)

## Motivation

Useful for altering response just before the response serialization (and validation) happens. The response schema is awesome as a response filter, but a response transformer (or anything, since this is an user-defined function) that runs just before the response schema validation would be great.

## Example

- Using a hashids preSerialization to convert integer resource ids on responses into hashids for some resource routes
- Outputs different level of visibility of the same resource depending on the authenticated client's privilege level who requested it. (Strips some object properties for ordinary users, strips none for admins, etc)
- Adding extra metadata on responses

### PR DESCRIPTION
**Pull request by ShogunPanda**
I've added preSerialization hook and route option. This fixes #1369.
One use-case for this is to being able to intercept serialization error before fast-json-stringify (which would hide all information since the error thrown has no context).

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Routes.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ They need to be in
 * `attachValidation`: attach `validationError` to request, if there is a schema validation error, instead of sending the error to the error handler.
 * `preValidation(request, reply, done)`: a [function](https://github.com/fastify/fastify/blob/master/docs/Hooks.md#route-hooks) called after the shared `preValidation` hooks, useful if you need to perform authentication at route level for example, it could also be and array of functions.
 * `preHandler(request, reply, done)`: a [function](https://github.com/fastify/fastify/blob/master/docs/Hooks.md#route-hooks) called just before the request handler, it could also be and array of functions.
+* `preSerialization(request, reply, payload, done)`: a [function](https://github.com/fastify/fastify/blob/master/docs/Hooks.md#route-hooks) called just before the serialization, it could also be and array of functions.
```
Comment (cemremengu) [inline docs/Routes.md]: There is a typo here repeated in others as well `and array of functions` -> `an array of functions`

Comment (ShogunPanda) [reply docs/Routes.md]: Sure, will fix tomorrow morning and then it will be ready to go :)

File: docs/Hooks.md [hunk 2]
```diff
@@ -169,6 +185,25 @@ fastify.addHook('onError', async (request, reply, error) => {
...
+
+```js
+fastify.addHook('preSerialization', (request, reply, payload, next) => {
+  var err = null;
+  var newPayload = {wrapped: payload }
+  next(err, newPayload)
+})
+
+// Or async
+fastify.addHook('onSend', async (request, reply, payload) => {
```
Comment (mcollina) [inline docs/Hooks.md]:
```suggestion
fastify.addHook('preSerialization', async (request, reply, payload) => {
```

Comment (ShogunPanda) [reply docs/Hooks.md]: Fixed, thanks!

File: docs/Routes.md (hunk 1 above)
Comment (ShogunPanda) [reply docs/Routes.md]: Fixed!

**Comment by ShogunPanda** [comment]
@mcollina Here's the results. I runned both of them twice after starting the server.

**Master**
```
Running 5s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 17 ms │ 18 ms │ 1.67 ms │ 5.17 ms │ 114.94 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 53151   │ 53151   │ 58815   │ 59519   │ 57699.2 │ 2357.83 │ 53151   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 8.72 MB │ 8.72 MB │ 9.65 MB │ 9.76 MB │ 9.46 MB │ 383 kB  │ 8.72 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

288k requests in 5.05s, 47.3 MB read
```

**PR Branch**
```
Running 5s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 16 ms │ 17 ms │ 1.64 ms │ 5.08 ms │ 131.24 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 55167   │ 55167   │ 60223   │ 60575   │ 58928   │ 2023.96 │ 55154   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 9.05 MB │ 9.05 MB │ 9.88 MB │ 9.94 MB │ 9.67 MB │ 332 kB  │ 9.05 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

295k requests in 5.04s, 48.3 MB read
```

**Comment by mcollina** [comment]
I wouldn’t expect this PR to be faster in any way. Can you confirm that?

**Comment by ShogunPanda** [comment]
I also updated typings, which I completedly forgot.

File: fastify.d.ts [hunk 3]
```diff
@@ -7,7 +7,6 @@
 import * as http from 'http'
 import * as http2 from 'http2'
 import * as https from 'https'
-import * as tls from 'tls'
```
Comment (mcollina) [inline fastify.d.ts]: why was this removed?

File: fastify.d.ts [hunk 4]
```diff
@@ -529,6 +547,12 @@ declare namespace fastify {
      */
     addHook(name: 'preValidation', hook: FastifyMiddleware<HttpServer, HttpRequest, HttpResponse>): FastifyInstance<HttpServer, HttpRequest, HttpResponse>
 
+    /**
+     * Hook that is fired after a request is processed, but before the response is serialized
+     * hook
+     */
+    addHook(name: 'preSerialization', hook: (this: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, req: FastifyRequest<HttpRequest>, reply: FastifyReply<HttpResponse>, payload: any, done: (err?: Error, value?: any) => void) => void): FastifyInstance<HttpServer, HttpRequest, HttpResponse>
```
Comment (mcollina) [inline fastify.d.ts]: I think this should be the same type that is passed to the preSerialization option.

File: fastify.d.ts (hunk 3 above)
Comment (ShogunPanda) [reply fastify.d.ts]: I think it was not used in the file and my editor removed it automatically. Is there a reason to keep it?

File: fastify.d.ts (hunk 4 above)
Comment (ShogunPanda) [reply fastify.d.ts]: Good point. Changed.

Comment (mcollina) [reply fastify.d.ts]: I cannot see any changes here.

Comment (ShogunPanda) [reply fastify.d.ts]: Ehm... I mean "I have it changed it locally and it will be there in my next push". Sorry for the confusion.

Comment (ShogunPanda) [reply fastify.d.ts]: No the commits are there.

**Comment by ShogunPanda** [comment]
Types tests added.

File: test/types/index.ts [hunk 5]
```diff
@@ -127,6 +129,12 @@ server.addHook('onResponse', function (req, reply, next) {
   }, 100)
 })
 
+server.addHook('preSerialization', function (req, reply, payload, next) {
+  this.log.debug('`this` is not `any`')
+  console.log(`${req.req.method} ${req.req.url}`)
+  next(undefined, payload)
+})
+
```
Comment (mcollina) [inline test/types/index.ts]: I think TS is missing a test for the new option to  `route`.

Comment (ShogunPanda) [reply test/types/index.ts]: You're right. I added it now.
